### ISSUE TITLE
custom JdbcEventListeners and filtering

### ISSUE THREAD

**Issue by jorgheymans** [p6spy/p6spy#418]
Over at the brave project (https://github.com/openzipkin/brave/blob/master/instrumentation/p6spy/src/main/java/brave/p6spy/TracingJdbcEventListener.java), there is an implementation of JdbcEventListener that forwards p6spy log events to zipkin. I noticed that executionThreshold was not respected in this implementation, events were being recorded regardless their execution time. Digging around the code i realized that in fact none of the filter configuration that p6spy offers apply in a custom JdbcEventListener, since the filtering is done in P6LogQuery which does not expose any of the decision logic.

So i am wondering if and how a custom jdbceventlistener can benefit from the same filtering mechanisms as the logging listeners do ?

**Comment by felixbarny** [comment]
Currently, they just can't because these are completely separate modules. P6LogQuery is something that is implemented on top of the JdbcEventListener abstraction.

**Comment by typekpb** [comment]
well, honestly I believe it's OK the way we did it, right? You might want to log respecting log levels, but on the other hand, you might want to process events for the completely different purpose.

@jorgheymans @felixbarny do we need updates in the p6spy here, or can we live with the current behaviour?

**Comment by jorgheymans** [comment]
IMO it can be surprising, from an external implementation point of view, that none of the filter configuration applies as soon as you implement your own listener. It made me look for a higher level extension point (which led me to P6LogQuery), because it seemed a bit like brave had chosen the wrong interface to implement.

**Comment by felixbarny** [comment]
You definitely chose the right interface. Actually, the logging stuff isn't part if p6spy's public API, although, admittedly not stated anywhere. In contrast, the event listener API was designed to be used by external projects.

It should be quite trivial to implement similar filtering capabilities in brave.

I don't think it would make sense to make the filtering part reusable because of increased coupling and because it's actually quite trivial.

**Comment by typekpb** [comment]
sounds like we should close this one with `won't fix` label.
objections, anyone?

**Comment by jorgheymans** [comment]
yep there is nothing to do here really, other than maybe clarify a bit in the docs. It's a very niche issue anyway, don't think there are that many custom eventlistener impls out there. Thanks for the feedback!

**Comment by typekpb** [comment]
@jorgheymans excelent, thanks for feedback!

**Comment by typekpb** [comment]
I've seen this issue mentioned in the: https://github.com/openzipkin/brave/issues/824 so it obviously is a pain point. Maybe we should expose filtering capabilities in a common API usable by anyone (core as well as custom listeners)

**Comment by gavlyukovskiy** [comment]
I went through it one more time and I don't see how this filtering could help brave in particular or anyone else in general. 
What should be the behavior of this filtering mechanism? Should this logic apply to all listeners or you should be able to define concrete listeners where logic applies (e.g. I want to see only slow queries in the logs, but I need other listeners to get all callbacks)? Should this filtering logic apply only to query callbacks, but not connection/result set callbacks?

In brave there are two callbacks in use - `onBeforeAnyExecute` and `onAfterAnyExecute`. With filtering mechanism in place what do we expect?
1. `onBeforeAnyExecute` to be called for every query, but `onAfterAnyExecute` only if threshold is exceeded
2. first wait threshold and then immediately call `onBeforeAnyExecute` and `onAfterAnyExecute`

Both variants would break logic inside this particular listener, in first case span would be created but not closed and in second case span duration would be 0. 

Instead, since brave supports creating spans with timestamp in the past, it can be easily implemented using only `onAfterAnyExecute` callback with following logic:
```
public void onAfterAnyExecute(StatementInformation info, long elapsed, SQLException e) {
    if (elapsed > executionThreshold) {
        ...
        span.start(tracing.clock(span.context()).currentTimeMicroseconds() - (elapsed / 1000));
        span.finish();
    }
}
```
and it's the only way to achieve this - check after query execution if threshold is exceeded or not. As a convenient way to do it p6spy could introduce `SlowQueryListener` with a single abstract method `onAfterSlowQuery`.
